core/src/main/java/org/apache/accumulo/core/conf/AccumuloConfiguration.java

```diff
@@ -435,6 +438,60 @@ Integer getDeprecatedScanThreads(String name) {
     return null;
   }
 
+  private static class RefCount<T> {
+    T obj;
+    long count;
+
+    RefCount(long c, T r) {
+      this.count = c;
+      this.obj = r;
+    }
+  }
+
+  private class Deriver<T> implements Supplier<T> {
+
+    private AtomicReference<RefCount<T>> refref;
+    private Function<AccumuloConfiguration,T> factory;
+
+    Deriver(Function<AccumuloConfiguration,T> factory) {
+      this.factory = factory;
+      refref = new AtomicReference<>();
+    }
+
+    @Override
+    public T get() {
```

Comment:
This whole implementation could alternatively be:

```java
public T get() {
  return refref.updateAndGet(rc -> {
    long uc = getUpdateCount();
    // check if this is the first time getting the object, or if the config changed
    if (rc != null && rc.count == uc) {
      return rc; // nothing changed
    }
    // config changed, re-generate
    return new RefCount<>(uc, factory.apply(AccumuloConfiguration.this));
  }).obj;
}
```

I know you tend to have a preference for the `compareAndSet` method directly, but I think it's too low-level an operation to be intuitive to many developers. I much prefer the use of Java 8's new `updateAndGet` method. This also has the nice behavior of retrying (regenerating the object) if the config changed while we were executing, which is something I suggested doing in my other comment, since the result of `compareAndSet` wasn't being checked. Config changes are rare, so this should almost never block, and when it does, the result will always be generated from the most current configuration, and we'll throw away any stale copies (which is what we'd do anyway in my other comment's `do...while` suggestion).

Reply:
updateAndGet calls compare and set in a loop, which I would like to avoid.  I added comments to the code try to explain its goals and the thoughts behind its design.  Let me know what you think of the comments.